### ISSUE TITLE
Error in building in NativeScript Ng2 App - Failed to find Module

### ISSUE THREAD

**Issue by kmuthukrishnan** [NativeScript/NativeScript#3858]
Hi, 

I am trying to build a native script ng2 project using algolia search and algolia search helper from npm. These node modules appear to have a tty dependency.

### Tell us about the problem
When I try to run the project I am getting the following error:

Error: com.tns.NativeScriptException: Failed to find module: "tty", relative to: app/tns_modules/
(Full log below)

### Did you verify this is a real problem by searching [Stack Overflow](http://stackoverflow.com/questions/tagged/nativescript) and the [other open issues in this repo](https://github.com/NativeScript/nativescript/issues)?

Yes and I tried the following with no success:
a. delete npm, platform and hooks and reinstall
b. checked whether tty dependency is in package.json (it is)

Please suggest what I could try.

Search is very important to our app and alogolia is preferred. 

### Which platform(s) does your issue occur on?
Android (haven't tried on iOS yet)

### Please provide the following version numbers that your issue occurs with:
- CLI: 2.5.0
- Cross-platform modules: 2.5.0
- Runtime(s): "tns-android": version 2.5.0
- Plugin(s): 
package.json dependencies are:
"dependencies": {
    "@angular/common": "2.4.3",
    "@angular/compiler": "2.4.3",
    "@angular/core": "2.4.3",
    "@angular/forms": "2.4.3",
    "@angular/http": "2.4.3",
    "@angular/platform-browser": "2.4.3",
    "@angular/platform-browser-dynamic": "2.4.3",
    "@angular/router": "3.4.3"
    "@types/algoliasearch": "^3.18.0",
    "algoliasearch": "^3.22.1",
    "nativescript-angular": "1.4.0",
    "nativescript-theme-core": "~1.0.2",
    "reflect-metadata": "~0.1.8",
    "rxjs": "~5.0.1",
    "tns-core-modules": "2.5.0",
    "tty": "^1.0.1"
  },
  "devDependencies": {
    "babel-traverse": "6.4.5",
    "babel-types": "6.4.5",
    "babylon": "6.4.5",
    "lazy": "1.0.11",
    "nativescript-dev-android-snapshot": "^0.*.*",
    "nativescript-dev-typescript": "^0.3.5",
    "typescript": "~2.1.0",
    "zone.js": "~0.7.2"
  }

### Error log below
```
`An uncaught Exception occurred on "main" thread.
java.lang.RuntimeException: Unable to create application com.tns.NativeScriptApplication: com.tns.NativeScriptException: 

Error calling module function 
Error calling module function 
Error calling module function 
Error calling module function 
Error calling module function 
Error calling module function 
Error calling module function 
Error calling module function

Error: com.tns.NativeScriptException: Failed to find module: "tty", relative to: app/tns_modules/
    com.tns.Module.resolvePathHelper(Module.java:159)
    com.tns.Module.resolvePath(Module.java:60)
    com.tns.Runtime.runModule(Native Method)
    com.tns.Runtime.runModule(Runtime.java:580)
    com.tns.Runtime.run(Runtime.java:574)
    com.tns.NativeScriptApplication.onCreate(NativeScriptApplication.java:17)
    android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1014)
    android.app.ActivityThread.handleBindApplication(ActivityThread.java:4726)
    android.app.ActivityThread.access$1600(ActivityThread.java:153)
    android.app.ActivityThread$H.handleMessage(ActivityThread.java:1412)
    android.os.Handler.dispatchMessage(Handler.java:102)
    android.os.Looper.loop(Looper.java:148)
    android.app.ActivityThread.main(ActivityThread.java:5438)
    java.lang.reflect.Method.invoke(Native Method)
    com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:739)
    com.android.internal.os.ZygoteInit.main(ZygoteInit.java:629)
File: "<unknown>, line: 1, column: 265
```

```
StackTrace: 
```
```

            Frame: function:'require', file:'', line: 1, column: 266
            Frame: function:'', file:'file:///data/data/org.nativescript.Testing/files/app/tns_modules/debug/node.js', line: 6, column: 11
            Frame: function:'require', file:'', line: 1, column: 266
            Frame: function:'', file:'file:///data/data/org.nativescript.Testing/files/app/tns_modules/debug/index.js', line: 9, column: 20
            Frame: function:'require', file:'', line: 1, column: 266
            Frame: function:'', file:'file:///data/data/org.nativescript.Testing/files/app/tns_modules/algoliasearch/src/server/builds/node.js', line: 6, column: 13
            Frame: function:'require', file:'', line: 1, column: 266
            Frame: function:'', file:'file:///data/data/org.nativescript.Testing/files/app/tns_modules/algoliasearch/index.js', line: 6, column: 18
            Frame: function:'require', file:'', line: 1, column: 266
            Frame: function:'', file:'file:///data/data/org.nativescript.Testing/files/app/services/algolia.service.js', line: 3, column: 21
            Frame: function:'require', file:'', line: 1, column: 266
            Frame: function:'', file:'file:///data/data/org.nativescript.Testing/files/app/app.component.js', line: 3, column: 25
            Frame: function:'require', file:'', line: 1, column: 266
            Frame: function:'', file:'file:///data/data/org.nativescript.Testing/files/app/app.module.js', line: 4, column: 23
            Frame: function:'require', file:'', line: 1, column: 266
            Frame: function:'', file:'file:///data/data/org.nativescript.Testing/files/app/main.js', line: 3, column: 20
            Frame: function:'require', file:'', line: 1, column: 266

Error: com.tns.NativeScriptException: Failed to find module: "tty", relative to: app/tns_modules/
    com.tns.Module.resolvePathHelper(Module.java:159)
    com.tns.Module.resolvePath(Module.java:60)
    com.tns.Runtime.runModule(Native Method)
    com.tns.Runtime.runModule(Runtime.java:580)
    com.tns.Runtime.run(Runtime.java:574)
    com.tns.NativeScriptApplication.onCreate(NativeScriptApplication.java:17)
    android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1014)
    android.app.ActivityThread.handleBindApplication(ActivityThread.java:4726)
    android.app.ActivityThread.access$1600(ActivityThread.java:153)
    android.app.ActivityThread$H.handleMessage(ActivityThread.java:1412)
    android.os.Handler.dispatchMessage(Handler.java:102)
    android.os.Looper.loop(Looper.java:148)
    android.app.ActivityThread.main(ActivityThread.java:5438)
    java.lang.reflect.Method.invoke(Native Method)
    com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:739)
    com.android.internal.os.ZygoteInit.main(ZygoteInit.java:629)
File: "<unknown>, line: 1, column: 265

StackTrace: 
           Frame: function:'require', file:'', line: 1, column: 266
            Frame: function:'', file:'file:///data/data/org.nativescript.Testing/files/app/tns_modules/debug/node.js', line: 6, column: 11
            Frame: function:'require', file:'', line: 1, column: 266
            Frame: function:'', file:'file:///data/data/org.nativescript.Testing/files/app/tns_modules/debug/index.js', line: 9, column: 20
            Frame: function:'require', file:'', line: 1, column: 266
            Frame: function:'', file:'file:///data/data/org.nativescript.Testing/files/app/tns_modules/algoliasearch/src/server/builds/node.js', line: 6, column: 13
            Frame: function:'require', file:'', line: 1, column: 266
           Frame: function:'', file:'file:///data/data/org.nativescript.Testing/files/app/tns_modules/algoliasearch/index.js', line: 6, column: 18
            Frame: function:'require', file:'', line: 1, column: 266
           Frame: function:'', file:'file:///data/data/org.nativescript.Testing/files/app/services/algolia.service.js', line: 3, column: 21
            Frame: function:'require', file:'', line: 1, column: 266
            Frame: function:'', file:'file:///data/data/org.nativescript.Testing/files/app/app.component.js', line: 3, column: 25
            Frame: function:'require', file:'', line: 1, column: 266
            Frame: function:'', file:'file:///data/data/org.nativescript.Testing/files/app/app.module.js', line: 4, column: 23
            Frame: function:'require', file:'', line: 1, column: 266
            Frame: function:'', file:'file:///data/data/org.nativescript.Testing/files/app/main.js', line: 3, column: 20
            Frame: function:'require', file:'', line: 1, column: 266
```
<**The above trace keeps repeating>**

**Comment by tsonevn** [comment]
Hi @kmuthukrishnan,

Could you verify, whether this issue happens on clear `NativeScirpt Angular 2` project or on existing one? 
Something more, you could verify, if you have installed `tty` plugin as a global dependency while using `npm list -g` command. If so remove `tty` plugin with `npm uninstall tty -g` and build again the project.

Let me know, whether this helps

**Comment by vvo** [comment]
The problem is that the tty module is no more in npm: https://www.npmjs.com/package/tty, this one should be used as a "polyfill" to make it easy to port existing JavaScript modules to nativescript: https://github.com/substack/tty-browserify

But then you will have to somehow make it so that require("tty") is aliased or read as if the code was https://github.com/substack/tty-browserify. React Native had the same issue: https://github.com/facebook/react-native/issues/10

Basically the packager from Native Script should read the "browser" field of packages just like React Native does to mitigate those issues. Unsure there's already an issue about it.

**Comment by anuragd7** [comment]
@tsonevn 

Thanks for your suggestion. The issue was happening on an existing project. I built a new one and it is happening there too. 

I checked npm list -g - there was no tty plug in. Any other suggestions?

**Comment by anuragd7** [comment]
@vvo 
Thanks for your help. I am quite new to this, can you help me out with some instructions on how to proceed with the "polyfill".

Also, would something like [this](https://github.com/EddyVerbruggen/nativescript-nodeify) help?

**Comment by vvo** [comment]
@anuragd7 This issue would be better be solved at the native script module level. nodeify is not talking about tty.

**Comment by vvo** [comment]
@NickIliev Is this somehow solved?

I see those are relative issues:
- https://github.com/NativeScript/NativeScript/issues/195
- https://nativescript.ideas.aha.io/ideas/NS-I-30

**Comment by NickIliev** [comment]
Simply removing tty from the package.json file posted by @kmuthukrishnan is resolving this particular problem. The support for tty is logged as a feature [here ](https://github.com/NativeScript/NativeScript/issues/195)(as you have noticed) so I am closing this one as resolved and also as duplicate to the feature request.

Note: currently tty is not required by NativeScript as dependency and creating new Angular based project from scratch is not reproducing this behaviour.
